kernel/src/driver.rs

```diff
@@ -149,6 +150,16 @@ impl CommandResult {
     }
 }
 
+impl From<ReturnCode> for CommandResult {
```

Comment:
This is a bit tricky. One of the requirements of commands in 2.0 is that they have a single failure return type and a single success return type. Using this `From` means one could easily inadvertently violate this requirement. I'm worried about this because there are HILs which don't specify whether `ReturnCode::Success` or `ReturnCode::SuccessWithValue` is returned, so there could be implementations return both in different circumstances.

Reply:
Ah, I didn't realize that was a requirement -- just to clarify, for a given driver there can be multiple success types returned by a command, but for a given command_num only a single return type is possible? 

e.g. LedDriver::command() can return both `CommandResult::success()` and `CommandResult::success_u32()`, but only for different command numbers?

Reply:
I still feel it is important to have something like this in the kernel, rather than every driver have to reimplement this same logic -- there should only be one correct way to convert a `ReturnCode` to a `CommandResult`, depending on the allowable type of `ReturnCode` for the given command_num. 

It seems odd to me that otherwise every single capsule should have to reimplement the logic in your `console` example, for instance. How about adding the following to `CommandResult`?

```rust
    /// Converts a `ReturnCode` to either `CommandResult::Success` or `CommandResult::Failure`.
    /// Used to ensure that only a single success type can be returned from a given
    /// `command_num`. If `ReturnCode::SuccessWithValue` is passed, the value will be dropped.
    pub fn from_return_code_no_success_with_value(rc: ReturnCode) -> Self {
        match rc {
            ReturnCode::SuccessWithValue { .. } => Self::success(),
            ReturnCode::SUCCESS => Self::success(),
            _ => Self::failure(ErrorCode::try_from(rc).unwrap()),
        }
    }

    /// Converts a `ReturnCode` to either `CommandResult::SuccessU32` or `CommandResult::Failure`.
    /// Used to ensure that only a single success type can be returned from a given
    /// `command_num`. If ReturnCode::Success is passed, this function will panic.
    pub fn from_return_code_success_only_with_value(rc: ReturnCode) -> Self {
        match rc {
            ReturnCode::SuccessWithValue { value } => Self::success_u32(value as u32),
            ReturnCode::SUCCESS => panic!("Cannot create CommandResult"),
            _ => Self::failure(ErrorCode::try_from(rc).unwrap()),
        }
    }
```

Reply:
I'm wary of these wrappers as they abstract a vital part of encoding & communicating successes and errors to the userspace process, which is then no longer explicit but hidden behind a wrapper.

Really the only reason why we have this mechanical conversion to `CommandResult` at the end (which subjectively speaking I find somewhat misleading) is that we are currently in Stage 1 - so just mechanically porting capsules to the new API. In a Stage 2, we would be optimizing these capsules to make full use of the new ABI, so essentially returning a fully built `CommandResult` from each match arm. E.g. a "get current time" will have a completely different method of encoding it's return value to userspace than a "reset timer" (using different variants of the `CommandResult`).

It really doesn't make any sense to take two vastly independent operations (such as two different command numbers) and try to apply common operations on their return values, if that works it's going to be a coincidence.

Reply:
I don't want to use these to apply common operations on the return values from different command numbers, I want to use these to convert values within command numbers (I realize the example in this PR does not reflect that, I suggested these after learning about the new restrictions). Take the following example, from the analog comparator:

```rust
        match command_num {
            1 => match ErrorCode::try_from(self.comparison(channel)) {
                Err(_) => CommandResult::success(),
                Ok(e) => CommandResult::failure(e),
            },

            2 => match ErrorCode::try_from(self.start_comparing(channel)) {
                Err(_) => CommandResult::success(),
                Ok(e) => CommandResult::failure(e),
            },

            3 => match ErrorCode::try_from(self.stop_comparing(channel)) {
                Err(_) => CommandResult::success(),
                Ok(e) => CommandResult::failure(e),
            },
    }
```

```rust
        match command_num {
            1 => ErrorCode::from_return_code_no_success_with_value(self.comparison(channel)),

            2 => ErrorCode::from_return_code_no_success_with_value(self.start_comparing(channel)),

            3 => ErrorCode::from_return_code_no_success_with_value(self.stop_comparing(channel)),
        }
```

These do the same thing - wrap a subcommand where it is expected that the `ReturnCode` will not be `SuccessWithValue`. But in my opinion the second make it much more explicit what is happening, and has a lot less repeated boilerplate.

We don't yet have an example of the appropriate way to handle a ReturnCode that should be `SuccessWithValue` but ends up being `Success`, so we can a) let each capsule author decide what to do (make up a value? panic?) or we can provide a helper function that makes the decision uniform. 

I understand that in stage 2 this will not be necessary, but I would not be surprised if many capsules are never udpated to "stage 2", or not updated for a very long time.

Reply:
Only 16 capsules use `SuccessWithValue`, and at least a few of those only exist in command=0 to return like the number of LEDs or something. I'm not sure it makes sense to add all of this boilerplate:

```
            1 => match ErrorCode::try_from(self.comparison(channel)) {
                Err(_) => CommandResult::success(),
                Ok(e) => CommandResult::failure(e),
            },
```
for the relatively rare case, particularly when that boilerplate is quite confusing (an Err returns success??). It also makes the capsule hard to read, since the actual operation (doing a comparison) is buried in a lot of irrelevant text that a capsule reader doesn't need to understand.

Doing:
```
1 => ErrorCode::from_return_code(self.comparison(channel)),
```
is much more straightforward. `from_return_code` can just always assume that SuccessWithValue is not used, and we can special case the conversion in the capsules where `SuccessWithValue` is used.

Reply:
*shrug* SuccessWithValue was a mistake, we're now paying a cost for it. :(

I agree that your code is much more straightforward, but it's also incorrect. What if the `ReturnCode` is a success? `ErrorCode` does not encode success. It's an error code. This is a security-focused embedded kernel: we should not sacrifice correctness for code brevity.

In the particular case I was bringing up, the ambiguity of under-specified APIs could easily propagate to the syscall API and cause capsules to unknowingly violate the syscall API specification. Not good. The right approach is to have syscall drivers clean up possibly sloppy code, not cause that sloppiness to spread to userspace.

That being said. one reasonable way forward is to cut out SuccessWithValue. We can transition the capsules/traits that use it to use an alternative approach. I 100% agree that the current approach is verbose and am in favor of finding a cleaner way. But we need to do so without introducing errors.

Reply:
I'm not convinced that my latest proposed solution (see a couple comments up) is worse in any aspect than the example in the updates console or temperature drivers. The code in each the console and temperature drivers, if presented with a SuccessWithValue, would silently transform the SuccessWithValue to Success -- this in no way ensures that whoever ported the code actually traced all the way back to verify that a SuccessWithValue in that location was impossible. I don't see why putting the same functionality behind a single function call, explicitly labeled "no_value", somehow makes this mistake more or less likely. But if we really wanna make sure the code is correct, perhaps we should panic if a SuccessWithValue is present -- this is much more likely to be noticed during testing!

All for getting rid of SuccessWithValue, maybe those are the HILs that should first be converted to Result.

Reply:
> I'm not convinced that my latest proposed solution (see a couple comments up) is worse in any aspect than the example in the updates console or temperature drivers.

Your latest proposed solution seems reasonable. The big issue, of course, are the `panic!` calls. IMO, what's important is that the caller (e.g., the system call driver capsule) explicitly states what it expects, rather than doing an oblivious transformation. While I think the current behavior and invocation approach is correct, I agree its ergonomics are not great. One of the goals of this process is to figure out what the best ergonomics are and write them down, so we can translate the porting guide into a "writing syscall drivers guide".

Reply:
I realize this PR isn't really the right place for this discussion, and that I am maybe repeating myself a little, so I'm sorry for dragging it on.

But IMO this:
```rust
                    let rcode = self.driver.read_temperature();
                    let eres = ErrorCode::try_from(rcode);
                    match eres {
                        Ok(ecode) => CommandResult::failure(ecode),
                        _ => CommandResult::success(),
                    }
```

does not make it very clear `read_temperature()` is not allowed to return `ReturnCode::SuccessWithValue`.

Reply:
There are actually 0 uses of `SuccessWithValue` outside of userspace drivers (if you count IPC and memop as userspace drivers), so I think once all userspace drivers using it are transitioned to the new syscall API, removing `SuccessWithValue` will be really easy.

IMO, we should just make removing `SuccessWithValue` a blocking item for Tock 2.0, which lets everyone not worry about the possibility of its presence when updating each driver (and lets us provide an intuitive / easy to use `CommandResult::from(ReturnCode)`). Looking at the uses, I believe that completely removing `SuccessWithValue` will take only a couple hours once all 16 drivers using it are updated to the new syscall interface.

Reply:
Ok, I fleshed out my proposal in #2252, which is hopefully a more appropriate place for continued discussion. Marking this conversation as resolved, as the relevant code is no longer present in this PR anyway.